File: Packages/WhoopStore/Sources/WhoopStore/WhoopDeepCapabilities.swift
import WhoopProtocol

/// Explicit status vocabulary for WHOOP deep-data substrates tracked by #761.
///
/// This is deliberately separate from `WhoopLiveCapabilities`, which answers a different question:
/// which high-level product metrics a paired WHOOP can contribute. These types describe the current
/// confidence/support state of lower-level data paths so UI and diagnostics can distinguish waiting for
/// data from unsupported or experimental functionality without inventing a value.
public enum WhoopDeepCapability: String, CaseIterable, Sendable {
    case heartRate
    case hrv
    case respiration
    case spo2
    case skinTemperature
    case rawDeepBuffers
    case rawImu
    case battery
    case bodyStatus
    case syncMode
}

public enum WhoopCapabilityState: String, Equatable, Sendable {
    case available
    case unsupported
    case experimental
    case needsData
}

/// Stable, non-clinical explanation for why a capability is in its current state.
public enum WhoopCapabilityReason: String, Equatable, Sendable {
    case observationPresent
    case awaitingUsableData
    case calibratedLiveSpo2Unavailable
    case gatedResearchCapture
    case gatedRealtimeResearchCapture
    case gatedOffloadResearchCapture
    case validatedTransport
}

public struct WhoopCapabilityStatus: Equatable, Sendable {
    public let capability: WhoopDeepCapability
    public let state: WhoopCapabilityState
    public let reason: WhoopCapabilityReason

    public init(
        capability: WhoopDeepCapability,
        state: WhoopCapabilityState,
        reason: WhoopCapabilityReason
    ) {
        self.capability = capability
        self.state = state
        self.reason = reason
    }
}

/// Pure WHOOP capability-status resolver. Twin of Android `WhoopDeepCapabilities`.
///
/// `family` is the already-resolved protocol family from `WhoopProtocol.DeviceFamily`; this layer never
/// reparses registry/model strings or guesses hardware generation. `Whoop5Variant` remains orthogonal and
/// is intentionally not needed by the current axes because 5.0 and MG share these transport states.
///
/// `observed` is explicit caller evidence that a usable value/sample exists for this device/session.
/// It may promote only `needsData -> available`. It can never promote an `unsupported` or
/// `experimental` substrate, which prevents a stray raw/candidate packet from becoming a product claim.
public enum WhoopDeepCapabilities {

    public static func status(
        for capability: WhoopDeepCapability,
        family: DeviceFamily? = nil,
        observed: Set<WhoopDeepCapability> = []
    ) -> WhoopCapabilityStatus {
        switch capability {
        case .spo2:
            // #548: NOOP cannot produce a calibrated SpO2 percentage from a live WHOOP path.
            // Raw/candidate bytes therefore never auto-promote this capability.
            return .init(
                capability: capability,
                state: .unsupported,
                reason: .calibratedLiveSpo2Unavailable
            )

        case .rawDeepBuffers, .rawImu:
            // Instrument-first research paths. A decoder/capture existing is not equivalent to a
            // generally available product substrate, and observing one does not change this state.
            return .init(
                capability: capability,
                state: .experimental,
                reason: researchCaptureReason(for: family)
            )

        case .syncMode:
            return .init(
                capability: capability,
                state: .available,
                reason: .validatedTransport
            )

        case .heartRate, .hrv, .respiration, .skinTemperature, .battery, .bodyStatus:
            if observed.contains(capability) {
                return .init(
                    capability: capability,
                    state: .available,
                    reason: .observationPresent
                )
            }
            return .init(
                capability: capability,
                state: .needsData,
                reason: .awaitingUsableData
            )
        }
    }

    /// Stable roadmap-order profile, suitable for diagnostics/UI adapters without a second source of truth.
    public static func profile(
        family: DeviceFamily? = nil,
        observed: Set<WhoopDeepCapability> = []
    ) -> [WhoopCapabilityStatus] {
        WhoopDeepCapability.allCases.map {
            status(for: $0, family: family, observed: observed)
        }
    }

    private static func researchCaptureReason(for family: DeviceFamily?) -> WhoopCapabilityReason {
        switch family {
        case .some(.whoop5):
            // 5/MG validated high-rate IMU is received through the gated historical/offload path (#423).
            return .gatedOffloadResearchCapture
        case .some(.whoop4):
            // 4.0 high-rate raw streams are realtime and intentionally gated/silenced by default (#423).
            return .gatedRealtimeResearchCapture
        case .none:
            // Unknown family stays unknown. Generation resolution belongs to WhoopProtocol.
            return .gatedResearchCapture
        }
    }
}

File: android/app/src/main/java/com/noop/data/WhoopDeepCapabilities.kt
package com.noop.data

import com.noop.protocol.DeviceFamily

/** Explicit status vocabulary for WHOOP deep-data substrates tracked by #761. */
enum class WhoopDeepCapability(val wireName: String) {
    HEART_RATE("heartRate"),
    HRV("hrv"),
    RESPIRATION("respiration"),
    SPO2("spo2"),
    SKIN_TEMPERATURE("skinTemperature"),
    RAW_DEEP_BUFFERS("rawDeepBuffers"),
    RAW_IMU("rawImu"),
    BATTERY("battery"),
    BODY_STATUS("bodyStatus"),
    SYNC_MODE("syncMode"),
}

enum class WhoopCapabilityState(val wireName: String) {
    AVAILABLE("available"),
    UNSUPPORTED("unsupported"),
    EXPERIMENTAL("experimental"),
    NEEDS_DATA("needsData"),
}

enum class WhoopCapabilityReason(val wireName: String) {
    OBSERVATION_PRESENT("observationPresent"),
    AWAITING_USABLE_DATA("awaitingUsableData"),
    CALIBRATED_LIVE_SPO2_UNAVAILABLE("calibratedLiveSpo2Unavailable"),
    GATED_RESEARCH_CAPTURE("gatedResearchCapture"),
    GATED_REALTIME_RESEARCH_CAPTURE("gatedRealtimeResearchCapture"),
    GATED_OFFLOAD_RESEARCH_CAPTURE("gatedOffloadResearchCapture"),
    VALIDATED_TRANSPORT("validatedTransport"),
}

data class WhoopCapabilityStatus(
    val capability: WhoopDeepCapability,
    val state: WhoopCapabilityState,
    val reason: WhoopCapabilityReason,
)

/**
 * Pure WHOOP capability-status resolver. Behavioral twin of Swift `WhoopDeepCapabilities`.
 *
 * This is separate from [WhoopLiveCapabilities]: that registry advertises high-level product metrics,
 * while this model explains the support/confidence state of deep-data substrates. [family] is the
 * already-resolved canonical protocol family; this layer never reparses model labels or guesses a
 * generation. [observed] is explicit caller evidence of usable data and may promote only
 * NEEDS_DATA -> AVAILABLE. It never promotes UNSUPPORTED or EXPERIMENTAL paths.
 */
object WhoopDeepCapabilities {

    fun status(
        capability: WhoopDeepCapability,
        family: DeviceFamily? = null,
        observed: Set<WhoopDeepCapability> = emptySet(),
    ): WhoopCapabilityStatus = when (capability) {
        WhoopDeepCapability.SPO2 -> WhoopCapabilityStatus(
            capability,
            WhoopCapabilityState.UNSUPPORTED,
            WhoopCapabilityReason.CALIBRATED_LIVE_SPO2_UNAVAILABLE,
        )

        WhoopDeepCapability.RAW_DEEP_BUFFERS,
        WhoopDeepCapability.RAW_IMU -> WhoopCapabilityStatus(
            capability,
            WhoopCapabilityState.EXPERIMENTAL,
            researchCaptureReason(family),
        )

        WhoopDeepCapability.SYNC_MODE -> WhoopCapabilityStatus(
            capability,
            WhoopCapabilityState.AVAILABLE,
            WhoopCapabilityReason.VALIDATED_TRANSPORT,
        )

        else -> if (capability in observed) {
            WhoopCapabilityStatus(
                capability,
                WhoopCapabilityState.AVAILABLE,
                WhoopCapabilityReason.OBSERVATION_PRESENT,
            )
        } else {
            WhoopCapabilityStatus(
                capability,
                WhoopCapabilityState.NEEDS_DATA,
                WhoopCapabilityReason.AWAITING_USABLE_DATA,
            )
        }
    }

    /** Stable roadmap-order profile, matching Swift CaseIterable declaration order. */
    fun profile(
        family: DeviceFamily? = null,
        observed: Set<WhoopDeepCapability> = emptySet(),
    ): List<WhoopCapabilityStatus> = WhoopDeepCapability.entries.map {
        status(it, family, observed)
    }

    private fun researchCaptureReason(family: DeviceFamily?): WhoopCapabilityReason = when (family) {
        DeviceFamily.WHOOP5 -> WhoopCapabilityReason.GATED_OFFLOAD_RESEARCH_CAPTURE
        DeviceFamily.WHOOP4 -> WhoopCapabilityReason.GATED_REALTIME_RESEARCH_CAPTURE
        null -> WhoopCapabilityReason.GATED_RESEARCH_CAPTURE
    }
}
